middle_end/flambda2/types/env/typing_env.ml

```diff
@@ -1257,6 +1260,53 @@ end = struct
       next_binding_time = env.next_binding_time
     }
 
+  let create_from_closure_conversion_approx
+      (symbols : _ Value_approximation.t Symbol.Map.t) =
+    let defined_symbols_without_equations =
+      Symbol.Map.keys symbols |> Symbol.Set.elements
+    in
+    let code_age_relation = Code_age_relation.empty in
+    let next_binding_time = Binding_time.earliest_var in
+    let rec type_from_approx approx =
+      match (approx : _ Value_approximation.t) with
+      | Value_unknown -> MTC.unknown Flambda_kind.value
+      | Block_approximation (fields, alloc_mode) ->
+        let fields = List.map type_from_approx (Array.to_list fields) in
+        MTC.immutable_block ~is_unique:false Tag.zero
+          ~field_kind:Flambda_kind.value ~fields (Or_unknown.Known alloc_mode)
+      | Closure_approximation (code_id, _code_opt) ->
+        let closure_id =
+          Closure_id.wrap
+            (Compilation_unit.get_current_exn ())
+            (Variable.create (Code_id.name code_id))
```

Comment:
Note: this horrible hack is going to cause problems later (the closure ID could show up in the free names of the typing env later, leading to errors during the offset computations). In this PR we don't use the free names of the generated typing env, and @Keryan-dev has a cleaner fix that keeps the closure ID in the approximation as part as another PR, so we're leaving it as it is for now.

Reply:
Addressed in #424